### PR TITLE
Adding token based authentication

### PR DESCRIPTION
**Pull request by burzum**

### CODE REVIEW

File: src/Authenticator/TokenAuthenticator.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    protected function _getToken(ServerRequestInterface $request)
+    {
+        $header = $this->config('header');
+        if (!empty($header)) {
+            return $request->getHeaderLine($header);
+        }
+
+        $param = $this->config('queryParam');
+        $queryParams = $request->getQueryParams();
+        if (!empty($queryParams[$param])) {
```
Comment (markstory) [inline src/Authenticator/TokenAuthenticator.php]: Should you isset() instead of empty()?

File: src/Identifier/TokenIdentifier.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     * @return false|EntityInterface
+     */
+    public function identify($data)
+    {
+        if (!isset($data['token'])) {
+            return false;
+        }
+
+        $tokenVerification = $this->config('tokenVerification');
+        if (is_callable($tokenVerification)) {
```
Comment (markstory) [inline src/Identifier/TokenIdentifier.php]: Should you make sure the callable isn't a string?

File: tests/TestCase/Authenticator/TokenAuthenticatorTest.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        );
+        $response = new Response('php://memory', 200, ['X-testing' => 'Yes']);
+
+        // Test without token
+        $tokenAuth = new TokenAuthenticator($identifiers, [
+            'queryParam' => 'token'
+        ]);
+        $result = $tokenAuth->authenticate($request, $response);
+        $this->assertInstanceOf('\Authentication\Result', $result);
+        $this->assertEquals(Result::FAILURE_OTHER, $result->getCode());
```
Comment (markstory) [inline tests/TestCase/Authenticator/TokenAuthenticatorTest.php]: Should these cases be separate tests?

File: src/Authenticator/TokenAuthenticator.php (hunk 1 above)
Comment (burzum) [inline src/Authenticator/TokenAuthenticator.php]: Isset will return true even if it's an empty string or false. Empty seems more useful to me in this case?

> Determine if a variable is set and is not NULL. 

http://php.net/manual/en/function.isset.php

File: src/Identifier/TokenIdentifier.php (hunk 2 above)
Comment (burzum) [inline src/Identifier/TokenIdentifier.php]: Check the new commit.

File: tests/TestCase/Identifier/TokenIdentifierTest.php [hunk 4]
```diff
@@ -40,6 +42,30 @@ public function testIdentify()
...
+
+                return false;
+            }
+        ]);
+
+        $result = $identifier->identify(['token' => 'not-larry']);
+        $this->assertFalse($result);
+
+        $result = $identifier->identify(['token' => 'larry']);
+        $this->assertInstanceOf('\Cake\Datasource\EntityInterface', $result);
```
Comment (burzum) [inline tests/TestCase/Identifier/TokenIdentifierTest.php]: Feel free to change it? :)

**Pull request merged**